Copyright enforcer fails with file type changed

When changing the file type of a file (E.g. moving a file and replacing it with a symlink) the git log contains a line that starts with `T`.

E.g.

```
T       hols/oci-devops/init/locals.tf
T       hols/oci-devops/init/terraform.tfvars
T       hols/oci-devops/init/variables.tf
M       hols/oci-devops/main/destroy.sh
M       hols/oci-devops/main/get.sh
T       hols/oci-devops/main/locals.tf
T       hols/oci-devops/main/terraform.tfvars
T       hols/oci-devops/main/variables.tf
```

Exception:

```
Caused by: io.helidon.build.maven.enforcer.EnforcerException: Could not parse line T	hols/oci-devops/init/locals.tf
    at io.helidon.build.maven.enforcer.GitCommands.gitOp (GitCommands.java:330)
    at io.helidon.build.maven.enforcer.GitCommands.gitTracked (GitCommands.java:97)
    at io.helidon.build.maven.enforcer.FileFinder.findFiles (FileFinder.java:89)
    at io.helidon.build.maven.enforcer.EnforcerMojo.execute (EnforcerMojo.java:189)
```

The git command executed by [GitCommands.gitTracked](https://github.com/helidon-io/helidon-build-tools/blob/7bd17f3c960d9d8ee4de56d4b879fe13e0b436c7/maven-plugins/enforcer-maven-plugin/src/main/java/io/helidon/build/maven/enforcer/GitCommands.java#L75) is:

```shell
git log --pretty=%cd --date=short --name-status --reverse
```

The help of `--name-status` shows this:
```
See the description of the --diff-filter option on what the status letters mean.
```

The help of `--diff-filter` shows this:
```
Select only files that are Added (A), Copied (C), Deleted (D), Modified (M), Renamed (R), have their type (i.e. regular file, symlink, submodule, ...) changed (T), are Unmerged (U), are Unknown (X),
```

The code that parses the output is here: https://github.com/helidon-io/helidon-build-tools/blob/7bd17f3c960d9d8ee4de56d4b879fe13e0b436c7/maven-plugins/enforcer-maven-plugin/src/main/java/io/helidon/build/maven/enforcer/GitCommands.java#L314

It's missing `T`, `U` and `X`.

--- a/common/common/src/main/java/io/helidon/build/common/FileUtils.java
+++ b/common/common/src/main/java/io/helidon/build/common/FileUtils.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2019, 2024 Oracle and/or its affiliates.
+ * Copyright (c) 2019, 2025 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -15,7 +15,9 @@
  */
 package io.helidon.build.common;
 
+import java.io.BufferedReader;
 import java.io.File;
+import java.io.FileReader;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
@@ -24,6 +26,7 @@ import java.net.MalformedURLException;
 import java.net.URI;
 import java.net.URISyntaxException;
 import java.net.URL;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.FileSystem;
 import java.nio.file.FileSystemAlreadyExistsException;
 import java.nio.file.FileSystems;
@@ -1137,6 +1140,23 @@ public final class FileUtils {
             return path.toUri().toURL();
         } catch (MalformedURLException e) {
             throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Check if file is a submodule.
+     *
+     * @param path tested file path
+     * @return {@code true} if the file is a submodule, {@code false} otherwise
+     */
+    public static boolean isSubModule(Path path) {
+        try (BufferedReader reader = new BufferedReader(new FileReader(path.toFile(), StandardCharsets.UTF_8))) {
+            String line = reader.readLine();
+            return line == null
+                    ? false
+                    : line.startsWith("[submodule");
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
         }
     }
 }

--- a/common/common/src/main/java/io/helidon/build/common/FileUtils.java
+++ b/common/common/src/main/java/io/helidon/build/common/FileUtils.java
@@ -15,9 +15,7 @@
  */
 package io.helidon.build.common;
 
-import java.io.BufferedReader;
 import java.io.File;
-import java.io.FileReader;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
@@ -26,7 +24,6 @@ import java.net.MalformedURLException;
 import java.net.URI;
 import java.net.URISyntaxException;
 import java.net.URL;
-import java.nio.charset.StandardCharsets;
 import java.nio.file.FileSystem;
 import java.nio.file.FileSystemAlreadyExistsException;
 import java.nio.file.FileSystems;
@@ -1144,17 +1141,15 @@ public final class FileUtils {
     }
 
     /**
-     * Check if file is a submodule.
+     * Check if a line of that file matches predicate.
      *
-     * @param path tested file path
-     * @return {@code true} if the file is a submodule, {@code false} otherwise
+     * @param path file path
+     * @param predicate predicate
+     * @return {@code true} if matches, {@code false} otherwise
      */
-    public static boolean isSubModule(Path path) {
-        try (BufferedReader reader = new BufferedReader(new FileReader(path.toFile(), StandardCharsets.UTF_8))) {
-            String line = reader.readLine();
-            return line == null
-                    ? false
-                    : line.startsWith("[submodule");
+    public static boolean containsLine(Path path, Predicate<String> predicate) {
+        try (Stream<String> lines = Files.lines(path)){
+            return lines.anyMatch(predicate);
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }

--- a/common/common/src/test/java/io/helidon/build/common/FileUtilsTest.java
+++ b/common/common/src/test/java/io/helidon/build/common/FileUtilsTest.java
@@ -32,13 +32,13 @@ import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.condition.EnabledOnOs;
 import org.junit.jupiter.api.condition.OS;
 
+import static io.helidon.build.common.FileUtils.containsLine;
 import static io.helidon.build.common.FileUtils.ensureFile;
 import static io.helidon.build.common.FileUtils.list;
 import static io.helidon.build.common.FileUtils.newZipFileSystem;
 import static io.helidon.build.common.FileUtils.unique;
 import static io.helidon.build.common.FileUtils.unzip;
 import static io.helidon.build.common.FileUtils.zip;
-import static io.helidon.build.common.FileUtils.isSubModule;
 import static io.helidon.build.common.Unchecked.unchecked;
 import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
 import static java.nio.file.attribute.PosixFilePermission.GROUP_EXECUTE;
@@ -158,13 +158,13 @@ class FileUtilsTest {
     }
 
     @Test
-    void testIsSubmodule() throws IOException {
+    void testContainsLine() throws IOException {
         Path blue = TestFiles.targetDir(FileUtilsTest.class).resolve("test-classes/vfs/blue");
         Path submodule = ensureFile(outputDir.resolve("submodule"));
         Files.write(submodule, "[submodule ...".getBytes());
 
-        assertThat(isSubModule(submodule), is(true));
-        assertThat(isSubModule(blue), is(false));
+        assertThat(containsLine(submodule, line -> line.startsWith("[submodule")), is(true));
+        assertThat(containsLine(blue, line -> false), is(false));
     }
 
     private static void readZipFileContent(Path zip, Consumer<Path> consumer) {

--- a/common/common/src/test/java/io/helidon/build/common/FileUtilsTest.java
+++ b/common/common/src/test/java/io/helidon/build/common/FileUtilsTest.java
@@ -159,12 +159,11 @@ class FileUtilsTest {
 
     @Test
     void testContainsLine() throws IOException {
-        Path blue = TestFiles.targetDir(FileUtilsTest.class).resolve("test-classes/vfs/blue");
         Path submodule = ensureFile(outputDir.resolve("submodule"));
         Files.write(submodule, "[submodule ...".getBytes());
 
         assertThat(containsLine(submodule, line -> line.startsWith("[submodule")), is(true));
-        assertThat(containsLine(blue, line -> false), is(false));
+        assertThat(containsLine(submodule, line -> false), is(false));
     }
 
     private static void readZipFileContent(Path zip, Consumer<Path> consumer) {

--- a/common/common/src/test/java/io/helidon/build/common/FileUtilsTest.java
+++ b/common/common/src/test/java/io/helidon/build/common/FileUtilsTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2023, 2024 Oracle and/or its affiliates.
+ * Copyright (c) 2023, 2025 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -32,11 +32,13 @@ import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.condition.EnabledOnOs;
 import org.junit.jupiter.api.condition.OS;
 
+import static io.helidon.build.common.FileUtils.ensureFile;
 import static io.helidon.build.common.FileUtils.list;
 import static io.helidon.build.common.FileUtils.newZipFileSystem;
 import static io.helidon.build.common.FileUtils.unique;
 import static io.helidon.build.common.FileUtils.unzip;
 import static io.helidon.build.common.FileUtils.zip;
+import static io.helidon.build.common.FileUtils.isSubModule;
 import static io.helidon.build.common.Unchecked.unchecked;
 import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
 import static java.nio.file.attribute.PosixFilePermission.GROUP_EXECUTE;
@@ -153,6 +155,16 @@ class FileUtilsTest {
         Path wd = unique(outputDir, "e2e-permissions");
         Path zipDir = createZipDirectory(wd);
         zipAndUnzip(zipDir);
+    }
+
+    @Test
+    void testIsSubmodule() throws IOException {
+        Path blue = TestFiles.targetDir(FileUtilsTest.class).resolve("test-classes/vfs/blue");
+        Path submodule = ensureFile(outputDir.resolve("submodule"));
+        Files.write(submodule, "[submodule ...".getBytes());
+
+        assertThat(isSubModule(submodule), is(true));
+        assertThat(isSubModule(blue), is(false));
     }
 
     private static void readZipFileContent(Path zip, Consumer<Path> consumer) {

--- a/maven-plugins/enforcer-maven-plugin/src/main/java/io/helidon/build/maven/enforcer/FileFinder.java
+++ b/maven-plugins/enforcer-maven-plugin/src/main/java/io/helidon/build/maven/enforcer/FileFinder.java
@@ -31,9 +31,9 @@ import java.util.List;
 import java.util.Set;
 import java.util.stream.Collectors;
 
-import io.helidon.build.common.FileUtils;
 import io.helidon.build.common.logging.Log;
 
+import static io.helidon.build.common.FileUtils.containsLine;
 import static io.helidon.build.maven.enforcer.GitIgnore.create;
 
 /**
@@ -177,7 +177,7 @@ public class FileFinder {
             return false;
         }
 
-        if (FileUtils.isSubModule(file.path())) {
+        if (containsLine(file.path(), line -> line.startsWith("[submodule"))) {
             Log.debug("File " + file.relativePath() + " is a submodule, ignoring.");
             return false;
         }

--- a/maven-plugins/enforcer-maven-plugin/src/main/java/io/helidon/build/maven/enforcer/FileFinder.java
+++ b/maven-plugins/enforcer-maven-plugin/src/main/java/io/helidon/build/maven/enforcer/FileFinder.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, 2023 Oracle and/or its affiliates.
+ * Copyright (c) 2021, 2025 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -31,6 +31,7 @@ import java.util.List;
 import java.util.Set;
 import java.util.stream.Collectors;
 
+import io.helidon.build.common.FileUtils;
 import io.helidon.build.common.logging.Log;
 
 import static io.helidon.build.maven.enforcer.GitIgnore.create;
@@ -168,6 +169,16 @@ public class FileFinder {
 
         if (renamedPaths.contains(file.path())) {
             Log.debug("File " + file.relativePath() + " has been renamed, ignoring.");
+            return false;
+        }
+
+        if (Files.isSymbolicLink(file.path())) {
+            Log.debug("File " + file.relativePath() + " is a symbolic link, ignoring.");
+            return false;
+        }
+
+        if (FileUtils.isSubModule(file.path())) {
+            Log.debug("File " + file.relativePath() + " is a submodule, ignoring.");
             return false;
         }
 

--- a/maven-plugins/enforcer-maven-plugin/src/main/java/io/helidon/build/maven/enforcer/GitCommands.java
+++ b/maven-plugins/enforcer-maven-plugin/src/main/java/io/helidon/build/maven/enforcer/GitCommands.java
@@ -18,6 +18,7 @@ package io.helidon.build.maven.enforcer;
 
 import java.io.BufferedReader;
 import java.io.IOException;
+import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -78,54 +79,7 @@ public final class GitCommands {
 
         Process process = startProcess(root, "log", "--pretty=%cd", "--date=short", "--name-status", "--reverse");
 
-        Map<String, Integer> fileToYear = new HashMap<>();
-        int lastYear = -1;
-
-        try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
-            String line;
-            while ((line = reader.readLine()) != null) {
-                if (line.isBlank()) {
-                    continue;
-                }
-                Matcher matcher = DATE_PATTERN.matcher(line);
-                if (matcher.matches()) {
-                    lastYear = Integer.parseInt(matcher.group(1));
-                    continue;
-                }
-                if (lastYear == -1) {
-                    throw new EnforcerException("Failed to parse output, expecting date to be present");
-                }
-
-                GitOperation gitOp = gitOp(line);
-                String relativePath = stripGitOp(line);
-
-                switch (gitOp) {
-                    case DELETE:
-                        fileToYear.remove(relativePath);
-                        break;
-                    case RENAME:
-                        rename(fileToYear, relativePath, lastYear);
-                        break;
-                    case COPY:
-                        copy(fileToYear, relativePath, lastYear);
-                        break;
-                    case UNKNOWN:
-                        Log.warn(String.format("Unknown git operation for file: %s, ignored.", root.resolve(relativePath)));
-                        break;
-                    case CHANGED:
-                    case UNMERGED:
-                    case ADD:
-                    case MODIFY:
-                    default:
-                        // any other type modifies the timestamp
-                        fileToYear.put(relativePath, lastYear);
-                        // do nothing, it is already in
-                        break;
-                }
-            }
-        } catch (IOException e) {
-            throw new EnforcerException("Failed to read output when getting tracked files", e);
-        }
+        Map<String, Integer> fileToYear = gitLog(root, process.getInputStream());
 
         waitFor(process, String.valueOf(List.of()));
 
@@ -343,6 +297,65 @@ public final class GitCommands {
             return GitOperation.UNKNOWN;
         }
         throw new EnforcerException("Could not parse line " + line);
+    }
+
+    /**
+     * Return all files read from inputStream.
+     *
+     * @param root root path
+     * @param inputStream git log process input
+     * @return file name with year
+     */
+    static Map<String, Integer> gitLog(Path root, InputStream inputStream) {
+        Map<String, Integer> fileToYear = new HashMap<>();
+        int lastYear = -1;
+
+        try (BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream))) {
+            String line;
+            while ((line = reader.readLine()) != null) {
+                if (line.isBlank()) {
+                    continue;
+                }
+                Matcher matcher = DATE_PATTERN.matcher(line);
+                if (matcher.matches()) {
+                    lastYear = Integer.parseInt(matcher.group(1));
+                    continue;
+                }
+                if (lastYear == -1) {
+                    throw new EnforcerException("Failed to parse output, expecting date to be present");
+                }
+
+                GitOperation gitOp = gitOp(line);
+                String relativePath = stripGitOp(line);
+
+                switch (gitOp) {
+                    case DELETE:
+                        fileToYear.remove(relativePath);
+                        break;
+                    case RENAME:
+                        rename(fileToYear, relativePath, lastYear);
+                        break;
+                    case COPY:
+                        copy(fileToYear, relativePath, lastYear);
+                        break;
+                    case UNKNOWN:
+                        Log.warn(String.format("Unknown git operation for file: %s, ignored.", root.resolve(relativePath)));
+                        break;
+                    case CHANGED:
+                    case UNMERGED:
+                    case ADD:
+                    case MODIFY:
+                    default:
+                        // any other type modifies the timestamp
+                        fileToYear.put(relativePath, lastYear);
+                        // do nothing, it is already in
+                        break;
+                }
+            }
+        } catch (IOException e) {
+            throw new EnforcerException("Failed to read output when getting tracked files", e);
+        }
+        return fileToYear;
     }
 
     private enum GitOperation {

--- a/maven-plugins/enforcer-maven-plugin/src/main/java/io/helidon/build/maven/enforcer/GitCommands.java
+++ b/maven-plugins/enforcer-maven-plugin/src/main/java/io/helidon/build/maven/enforcer/GitCommands.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, 2023 Oracle and/or its affiliates.
+ * Copyright (c) 2021, 2025 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -33,6 +33,8 @@ import java.util.function.Predicate;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
+
+import io.helidon.build.common.logging.Log;
 
 /**
  * Utility class for git commands.
@@ -68,7 +70,7 @@ public final class GitCommands {
      * Get all files in a directory tracked by the repository.
      * This may return files that were locally deleted.
      *
-     * @param root root of the repository
+     * @param root      root of the repository
      * @param checkPath path to check (within the repository)
      * @return list of files tracked within the checkPath
      */
@@ -98,22 +100,27 @@ public final class GitCommands {
                 String relativePath = stripGitOp(line);
 
                 switch (gitOp) {
-                case DELETE:
-                    fileToYear.remove(relativePath);
-                    break;
-                case RENAME:
-                    rename(fileToYear, relativePath, lastYear);
-                    break;
-                case COPY:
-                    copy(fileToYear, relativePath, lastYear);
-                    break;
-                case ADD:
-                case MODIFY:
-                default:
-                    // any other type modifies the timestamp
-                    fileToYear.put(relativePath, lastYear);
-                    // do nothing, it is already in
-                    break;
+                    case DELETE:
+                        fileToYear.remove(relativePath);
+                        break;
+                    case RENAME:
+                        rename(fileToYear, relativePath, lastYear);
+                        break;
+                    case COPY:
+                        copy(fileToYear, relativePath, lastYear);
+                        break;
+                    case UNKNOWN:
+                        Log.warn(String.format("Unknown git operation for file: %s, ignored.", root.resolve(relativePath)));
+                        break;
+                    case CHANGED:
+                    case UNMERGED:
+                    case ADD:
+                    case MODIFY:
+                    default:
+                        // any other type modifies the timestamp
+                        fileToYear.put(relativePath, lastYear);
+                        // do nothing, it is already in
+                        break;
                 }
             }
         } catch (IOException e) {
@@ -137,10 +144,10 @@ public final class GitCommands {
      * Files locally modified.
      * Ignores deleted files - the user of these methods must consider files that no longer exist to be locally deleted.
      *
-     * @param root the root directory
-     * @param checkPath directory of interest
+     * @param root        the root directory
+     * @param checkPath   directory of interest
      * @param currentYear current year
-     * @param renamed a list to which this method adds the original paths of files that were renamed
+     * @param renamed     a list to which this method adds the original paths of files that were renamed
      * @return set of files in the directory of interest that were locally modified
      */
     static Set<FileRequest> locallyModified(Path root, Path checkPath, String currentYear, List<Path> renamed) {
@@ -326,7 +333,15 @@ public final class GitCommands {
         if (line.startsWith("R")) {
             return GitOperation.RENAME;
         }
-
+        if (line.startsWith("T\t")) {
+            return GitOperation.CHANGED;
+        }
+        if (line.startsWith("U\t")) {
+            return GitOperation.UNMERGED;
+        }
+        if (line.startsWith("X\t")) {
+            return GitOperation.UNKNOWN;
+        }
         throw new EnforcerException("Could not parse line " + line);
     }
 
@@ -335,6 +350,9 @@ public final class GitCommands {
         DELETE,
         MODIFY,
         RENAME,
-        COPY
+        COPY,
+        CHANGED,
+        UNMERGED,
+        UNKNOWN
     }
 }

--- a/maven-plugins/enforcer-maven-plugin/src/test/java/io/helidon/build/maven/enforcer/GitCommandTest.java
+++ b/maven-plugins/enforcer-maven-plugin/src/test/java/io/helidon/build/maven/enforcer/GitCommandTest.java
@@ -1,0 +1,71 @@
+/*
+ * Copyright (c) 2025 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.build.maven.enforcer;
+
+import java.io.ByteArrayInputStream;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.InputStream;
+import java.nio.file.Path;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import io.helidon.build.common.logging.Log;
+
+import org.junit.jupiter.api.Test;
+
+import static io.helidon.build.common.test.utils.TestFiles.testResourcePath;
+import static io.helidon.build.maven.enforcer.GitCommands.gitLog;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+class GitCommandTest {
+
+    @Test
+    void testGitLog() throws FileNotFoundException {
+        int warnings = Log.warnings();
+        Iterator<String> expected = List.of(
+                "foo/bar", "bar/foo", "foo/foo4", "foo/foo3", "foo/foo2", "foo/foo1").iterator();
+        InputStream is = new FileInputStream(testResourcePath(GitCommandTest.class, "git.txt").toFile());
+        Map<String, Integer> files = gitLog(Path.of("/root"), is);
+
+        for (Map.Entry<String, Integer> entry : files.entrySet()) {
+            assertThat(entry.getKey(), is(expected.next()));
+            assertThat(entry.getValue(), is(2000));
+        }
+
+        assertThat(expected.hasNext(), is(false));
+        assertThat("Warning message must be logged", Log.warnings(), is(warnings + 1));
+    }
+
+    @Test
+    void testWrongGitOperation() {
+        String content = "0000-00-00" + System.lineSeparator() + "H";
+        EnforcerException exception = assertThrows(EnforcerException.class,
+                () -> gitLog(Path.of("/root"), new ByteArrayInputStream(content.getBytes())));
+        assertThat(exception.getMessage(), is("Could not parse line H"));
+    }
+
+    @Test
+    void testWrongDate() {
+        String content = "000-00-00";
+        EnforcerException exception = assertThrows(EnforcerException.class,
+                () -> gitLog(Path.of("/root"), new ByteArrayInputStream(content.getBytes())));
+        assertThat(exception.getMessage(), is("Failed to parse output, expecting date to be present"));
+    }
+}

--- a/maven-plugins/enforcer-maven-plugin/src/test/resources/git.txt
+++ b/maven-plugins/enforcer-maven-plugin/src/test/resources/git.txt
@@ -1,0 +1,23 @@
+2000-00-00
+R	foo/foo		foo/bar
+
+2000-00-00
+D	foo/foo
+
+2000-00-00
+A	foo/foo1
+
+2000-00-00
+T	foo/foo2
+
+2000-00-00
+M	foo/foo3
+
+2000-00-00
+C	foo/foo		bar/foo
+
+2000-00-00
+U	foo/foo4
+
+2000-00-00
+X	foo/foo5

